### ISSUE TITLE
Generic, collection and dictionary support in CustomTypeMappings

### ISSUE THREAD

**Issue by bcrosnier** [jburzynski/TypeGen#89]
Hi,

We have a serializer that converts some C# dictionaries to [Maps](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map), and some lists to [Sets](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set).
`CustomTypeMappings` doesn't seem to handle dictionaries (or any other generic type).

# Steps to reproduce
Compile the following and run dotnet-typegen:
```csharp
    public class MyClass1
    {
        public Dictionary<string, MyClass2> MyDictionary { get; set; }
    }
    public class MyClass2
    {
        public string MyNullableProperty { get; set; }
    }

    public class ModelAssemblyGenerationSpec : GenerationSpec
    {
        public override void OnBeforeGeneration(OnBeforeGenerationArgs args)
        {
            // "System.Collections.Generic.Dictionary`2"
            args.GeneratorOptions.CustomTypeMappings.Add(typeof(Dictionary<,>).FullName, "Map<>");

            AddClass<MyClass1>();
            AddClass<MyClass2>();
        }
    }
```

# Expected
```ts
import { MyClass2 } from "./my-class2";

export class MyClass1 {
    myDictionary: Map<string,MyClass2>;
}
```

# Actual
```ts
import { MyClass2 } from "./my-class2";

export class MyClass1 {
    myDictionary: { [key: string]: MyClass2; };
}
```

# Notes

Sure enough, `CustomTypeMappings` is only used for simple types at this point ([TypeGen.Core/Generator/Services/TypeService.cs#L44](https://github.com/jburzynski/TypeGen/blob/master/src/TypeGen/TypeGen.Core/Generator/Services/TypeService.cs#L44)).

Would you be interested in one or more pull requests to add dictionaries, collections, and improve generic types support ?

Cheers !

**Comment by jburzynski** [comment]
Hi @bcrosnier ,

thanks for reporting this.

Something similar was also raised some time ago: #85 (it's about KeyValuePair, so also generic types not handled by customTypeMappings).

If you'd like to make a pull request to fix customTypeMappings, that would be great!

I think the fix would mostly be about changing TypeService.cs - I can't think of any other files that have any significant logic for handling customTypeMappings.

To give a general overview of the "architecture" so that it's a bit easier to do the fix: the class that generates the files is TypeGen.Core.Generator.Generator - this class uses other service classes to do particular things. And that's pretty much it :) TypeGen.Cli is only a console app that uses the Generator class to generate the files.

I could merge the PR and do a release at the beginning of January. If you'd like to create a NuGet package with the changes from the PR to use the package locally, you can see "docs/Release steps.txt". Essentially it's about running 3 scripts in the root of the repository:

- version-update.ps1 [old version] [new version]
- build-publish.ps1
- nuget-update.ps1

Thanks again!
Cheers

**Comment by jburzynski** [comment]
Hi, just an update that this functionality should be available in version 2.4.9 (I've just uploaded it minutes ago).

If you try it, could you please let me know if it works? Thanks :)

**Comment by jburzynski** [comment]
Closed due to inactivity
